# Natural Language Processing Project

# Multi-Label Categorical Classification of English Quotes

<b>Rish (Rishabh) Kaushick</b> <br>
College of Engineering <br>
Northeastern University <br>
Toronto, ON <br>
kaushick.r@northeastern.edu

# Downloading The Dataset From Hugging Face

In [1]:
from huggingface_hub import hf_hub_download
import pandas as pd

REPO_ID = "Abirate/english_quotes"
FILENAME = "quotes.jsonl"

# since the data is in jsonl format, we read it using read_json() function with lines=True
dataset = pd.read_json(hf_hub_download(repo_id=REPO_ID, filename=FILENAME, repo_type="dataset"), lines=True)


In [2]:
dataset

quote  \
0        “Be yourself; everyone else is already taken.”   
1     “I'm selfish, impatient and a little insecure....   
2     “Two things are infinite: the universe and hum...   
3                      “So many books, so little time.”   
4     “A room without books is like a body without a...   
...                                                 ...   
2503  “Morality is simply the attitude we adopt towa...   
2504  “Don't aim at success. The more you aim at it ...   
2505  “In life, finding a voice is speaking and livi...   
2506  “Winter is the time for comfort, for good food...   
2507                      “Silence is so freaking loud”   

                     author                                               tags  
0               Oscar Wilde  [be-yourself, gilbert-perreira, honesty, inspi...  
1            Marilyn Monroe  [best, life, love, mistakes, out-of-control, t...  
2           Albert Einstein  [human-nature, humor, infinity, philosophy, sc...  
3               Frank Zappa                                     [books, humor]  
4     Marcus Tullius Cicero                              [books, simile, soul]  
...                     ...                                                ...  
2503           Oscar Wilde,                             [morality, philosophy]  
2504      Viktor E. Frankl,                               [happiness, success]  
2505           John Grisham                               [inspirational-life]  
2506          Edith Sitwell                            [comfort, home, winter]  
2507          Sarah Dessen,    [just-listen, loud, owen, sara-dessen, silence]  

[2508 rows x 3 columns]

# Exploratory Data Analysis

In [3]:
dataset.head(10)

quote                 author  \
0     “Be yourself; everyone else is already taken.”            Oscar Wilde   
1  “I'm selfish, impatient and a little insecure....         Marilyn Monroe   
2  “Two things are infinite: the universe and hum...        Albert Einstein   
3                   “So many books, so little time.”            Frank Zappa   
4  “A room without books is like a body without a...  Marcus Tullius Cicero   
5  “Be who you are and say what you feel, because...      Bernard M. Baruch   
6  “You've gotta dance like there's nobody watchi...      William W. Purkey   
7  “You know you're in love when you can't fall a...              Dr. Seuss   
8  “You only live once, but if you do it right, o...               Mae West   
9  “Be the change that you wish to see in the wor...         Mahatma Gandhi   

                                                tags  
0  [be-yourself, gilbert-perreira, honesty, inspi...  
1  [best, life, love, mistakes, out-of-control, t...  
2  [human-nature, humor, infinity, philosophy, sc...  
3                                     [books, humor]  
4                              [books, simile, soul]  
5  [ataraxy, be-yourself, confidence, fitting-in,...  
6  [dance, heaven, hurt, inspirational, life, lov...  
7                     [dreams, love, reality, sleep]  
8                                      [humor, life]  
9  [action, change, inspirational, philosophy, wish]

In [6]:
# setting random_state in the random sample, so that each time the same value is presented
dataset.sample(5, random_state=23)

quote             author  \
2062  “Why are you worrying about YOU-KNOW-WHO, when...      J.K. Rowling,   
1078  “I am not sure that I exist, actually. I am al...  Jorge Luis Borges   
1176  “If you're lonely when you're alone, you're in...   Jean-Paul Sartre   
414   “You educate a man; you educate a man. You edu...      Brigham Young   
508   “There is neither happiness nor misery in the ...    Alexandre Dumas   

                                   tags  
2062                            [humor]  
1078                             [life]  
1176             [loneliness, solitude]  
414   [education, feminism, men, women]  
508       [hope, inspirational, wisdom]

In [7]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2508 entries, 0 to 2507
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   quote   2508 non-null   object
 1   author  2508 non-null   object
 2   tags    2508 non-null   object
dtypes: object(3)
memory usage: 58.9+ KB


In [8]:
# Quotes by J.k. Rowling
dataset.loc[dataset['author'] == 'J.K. Rowling,']

quote         author  \
11    “If you want to know what a man's like, take a...  J.K. Rowling,   
31    “It is our choices, Harry, that show what we t...  J.K. Rowling,   
35    “It does not do to dwell on dreams and forget ...  J.K. Rowling,   
53          “I solemnly swear that I am up to no good.”  J.K. Rowling,   
68    “It takes a great deal of bravery to stand up ...  J.K. Rowling,   
...                                                 ...            ...   
2321              “Don't let the muggles get you down.”  J.K. Rowling,   
2342  “I'm going to keep going until I succeed — o...  J.K. Rowling,   
2401  “You think the dead we loved ever truly leave ...  J.K. Rowling,   
2472  “Hello, Harry" said George, beaming at him. "W...  J.K. Rowling,   
2478  “But Dumbledore says he doesn't care what they...  J.K. Rowling,   

                                      tags  
11               [character, sirius-black]  
31                    [abilities, choices]  
35                          [dreams, life]  
53                              [mischief]  
68    [albus-dumbledore, courage, friends]  
...                                    ...  
2321         [encouragement, muggles, wit]  
2342                        [harry-potter]  
2401                                [dead]  
2472                              [humour]  
2478  [celebrity, dumbledore, fame, funny]  

[74 rows x 3 columns]

## Exploring The Target Variable - 'tags'

In [9]:
# Understanding the 'tags' attribute:
dataset['tags']

0       [be-yourself, gilbert-perreira, honesty, inspi...
1       [best, life, love, mistakes, out-of-control, t...
2       [human-nature, humor, infinity, philosophy, sc...
3                                          [books, humor]
4                                   [books, simile, soul]
                              ...                        
2503                               [morality, philosophy]
2504                                 [happiness, success]
2505                                 [inspirational-life]
2506                              [comfort, home, winter]
2507      [just-listen, loud, owen, sara-dessen, silence]
Name: tags, Length: 2508, dtype: object

In [10]:
dataset['tags'][0]

['be-yourself',
 'gilbert-perreira',
 'honesty',
 'inspirational',
 'misattributed-oscar-wilde',
 'quote-investigator']

In [11]:
dataset['tags'][0][0]

'be-yourself'

In [12]:
dataset['tags'][0][3]

'inspirational'

From the above we can see that the values in the 'tags' attribute are stored as a list.

In [13]:
# Looking for how many unique values are present in the author column
len(dataset['author'].unique())

880

In [14]:
dataset['tags'].value_counts()

tags
[love]                                                                                                                                                                                                                          75
[inspirational]                                                                                                                                                                                                                 63
[humor]                                                                                                                                                                                                                         37
[life]                                                                                                                                                                                                                          34
[writing]                                                                              

The above does not help us much because one row could have multiple categories in the list. <br>
Therefore, in order to see how many unique tags are there, we might need to iterate through the entire pandas dataframe, and store them in a Python set.

In [15]:
# iterating through the values in the pandas dataframe
tags = dataset['tags']
print(tags)
print(type(tags))

0       [be-yourself, gilbert-perreira, honesty, inspi...
1       [best, life, love, mistakes, out-of-control, t...
2       [human-nature, humor, infinity, philosophy, sc...
3                                          [books, humor]
4                                   [books, simile, soul]
                              ...                        
2503                               [morality, philosophy]
2504                                 [happiness, success]
2505                                 [inspirational-life]
2506                              [comfort, home, winter]
2507      [just-listen, loud, owen, sara-dessen, silence]
Name: tags, Length: 2508, dtype: object
<class 'pandas.core.series.Series'>


In [16]:
# Iterting though the pandas Series object and adding each value in a Set
# The benefit of the set is that only the unique values will be added, therefore all duplicates will be removed
unique_tags = set()

# row_tags is the list of tags in each row
# we can see that sometimes a single row has multiple tags, therefore, we will nest another for loop inside which reads individual tags
for row_tags in tags:
    for tag in row_tags:
        unique_tags.add(tag)

print(unique_tags)

{'writing', 'mother', 'darkness', 'real-life', 'knowledge-power', 'true-family', 'carroll', 'at-first-sight', 'self-discovery', 'recovery', 'thoughtful', 'self-determination', 'chivalry', 'humor-heaven', 'stealth', 'pause', 'youth', 'only-way', 'just-listen', 'misquote', 'high-school', 'bibliophiles', 'sarah-j-maas', 'elitist', 'apocryphal', 'calvin-and-hobbes', 'depression-silence', 'rocks', 'gandalf', 'obsession', 'holden-caulfield', 'gender-stereotypes', 'family', 'animal-farm', 'cruelty', 'bella', 'salinger', 'jesper', 'talent', 'listening', 'christina', 'god-religion-house', 'rich-heart', 'good-reads', 'equality', 'mankind', 'intellect', 'make-a-difference', 'books-vacation', 'double-standards', 'birth', 'drunk', 'living', 'authentic-living', 'save', 'antipathy', 'food', 'will-s-letter-to-tessa', 'wrong', 'ron', 'spinning', 'touchstones', 'walking', 'those-who-matter', 'fear', 'history', 'assumptions', 'inspirational-faith', 'october', 'melancholy', 'crooked-kingdom', 'thought', '

In [17]:
len(unique_tags)

2161

There are a total of 2,161 tags. This is too many if we want this to be our target variable.

In [18]:
# Let me create a Python dictionary to store the tag and it's corresponding frequency throughout this dataset.
tag_freq = {}

for row_tag in tags:
    for tag in row_tag:
        if tag in tag_freq:
            tag_freq[tag] += 1
        else:
            tag_freq[tag] = 1

print(tag_freq)

{'be-yourself': 16, 'gilbert-perreira': 1, 'honesty': 12, 'inspirational': 319, 'misattributed-oscar-wilde': 2, 'quote-investigator': 1, 'best': 1, 'life': 295, 'love': 327, 'mistakes': 10, 'out-of-control': 1, 'truth': 61, 'worst': 1, 'human-nature': 9, 'humor': 254, 'infinity': 4, 'philosophy': 60, 'science': 26, 'stupidity': 7, 'universe': 6, 'books': 141, 'simile': 8, 'soul': 12, 'ataraxy': 4, 'confidence': 11, 'fitting-in': 1, 'individuality': 21, 'misattributed-dr-seuss': 1, 'those-who-matter': 1, 'dance': 5, 'heaven': 2, 'hurt': 8, 'sing': 2, 'dreams': 31, 'reality': 17, 'sleep': 7, 'action': 7, 'change': 34, 'wish': 2, 'character': 11, 'sirius-black': 2, 'friends': 28, 'friendship': 63, 'misattributed-albert-camus': 1, 'lies': 12, 'lying': 6, 'memory': 9, 'friend': 3, 'knowledge': 17, 'enemies': 4, 'forgiveness': 10, 'strategy': 3, 'carpe-diem': 7, 'education': 22, 'learning': 15, 'darkness': 7, 'drive-out': 1, 'hate': 10, 'light': 11, 'peace': 14, 'music': 31, 'intelligence': 

In [19]:
# Let us sort the dictionary in descending order of the frequencies:
tag_freq_sorted = sorted(tag_freq.items(), key=lambda x:x[1], reverse=True)
print(tag_freq_sorted)

[('love', 327), ('inspirational', 319), ('life', 295), ('humor', 254), ('books', 141), ('inspiration', 87), ('reading', 86), ('happiness', 68), ('writing', 67), ('inspirational-quotes', 65), ('friendship', 63), ('wisdom', 62), ('truth', 61), ('philosophy', 60), ('optimism', 48), ('living', 48), ('life-quotes', 48), ('death', 47), ('hope', 46), ('poetry', 45), ('romance', 45), ('positive-affirmation', 41), ('positive-thinking', 41), ('women', 40), ('life-lessons', 40), ('optimistic', 40), ('positive', 40), ('positive-life', 40), ('inspire', 38), ('inspiring', 38), ('courage', 37), ('life-and-living', 36), ('change', 34), ('pain', 34), ('funny', 34), ('inspirational-life', 34), ('relationships', 33), ('religion', 32), ('dreams', 31), ('music', 31), ('fear', 31), ('jace-wayland', 29), ('friends', 28), ('strength', 28), ('freedom', 28), ('god', 28), ('inspirational-attitude', 28), ('motivational', 27), ('loneliness', 27), ('science', 26), ('beauty', 26), ('time', 25), ('motivation', 25), (

In [20]:
# printing the top 40 most frequent tags
for i in range(0, len(tag_freq_sorted)):
    print(tag_freq_sorted[i])
    if i == 40:
        break


('love', 327)
('inspirational', 319)
('life', 295)
('humor', 254)
('books', 141)
('inspiration', 87)
('reading', 86)
('happiness', 68)
('writing', 67)
('inspirational-quotes', 65)
('friendship', 63)
('wisdom', 62)
('truth', 61)
('philosophy', 60)
('optimism', 48)
('living', 48)
('life-quotes', 48)
('death', 47)
('hope', 46)
('poetry', 45)
('romance', 45)
('positive-affirmation', 41)
('positive-thinking', 41)
('women', 40)
('life-lessons', 40)
('optimistic', 40)
('positive', 40)
('positive-life', 40)
('inspire', 38)
('inspiring', 38)
('courage', 37)
('life-and-living', 36)
('change', 34)
('pain', 34)
('funny', 34)
('inspirational-life', 34)
('relationships', 33)
('religion', 32)
('dreams', 31)
('music', 31)
('fear', 31)


From the above we can see two main issues with the tags as highlighted in blue and yellow:
1. We can see different columns which conceptually mean the same thing however it is written in different ways. For example, inspirational, inspiration, inspire and inspiring – all mean the same thing. However, it is classified as different tags. For fixing this issue, we must perform lemmatization on the target column.
2. We can see multiple different words separated by a hyphen. For example, positive-affirmation, life-lessons, etc. It will be better if we split these words with the hyphen as a delimiter and add the respective categories separately. For example, ‘positive-affirmation’ will be replaced with the tags positive and affirmation. Performing this first will further help correctly lemmatize the words.

# Data Cleaning
## Seperating Words That Contain a Hyphen

In [24]:
dataset2 = dataset
for i in range(1, len(dataset2)):
    for j in range(0, len(dataset2['tags'][i])):
        # in this loop we can go through individual tag elements
        if('-' in dataset2['tags'][i][j]):
            # here the word contains a '-' character, therefore we can split the data
            # print(dataset2['tags'][i][j])
            new_split_strings = str(dataset2['tags'][i][j]).split('-')
            dataset2['tags'][i].pop(j)
            for new_string in new_split_strings:
                # print(new_string)
                dataset2['tags'][i].append(new_string)
            # print(type(dataset2['tags'][i])
            # print("This is the string: "),
            # print(new_split_strings)
        # print(dataset2['tags'][i][j]),
    # print(dataset2['tags'][i])
    # below code was written just a safety measure
    # if (i == 10):
    #     break


In [25]:
# writing a function to calculate the count of tags

def calculate_count_of_tags(tags=dataset['tags']):
    tag_count_dict = {}
    for row in tags:
        for tag in row:
            if tag in tag_count_dict:
                tag_count_dict[tag] += 1
            else:
                tag_count_dict[tag] = 1
    # now sorting it based on the highest values
    tag_freq_sorted = sorted(tag_count_dict.items(), key=lambda x:x[1], reverse=True)
    return tag_freq_sorted
    

In [26]:
tag_freq_sorted = calculate_count_of_tags(tags=dataset2['tags'])

In [28]:
for i in range(0, len(tag_freq_sorted)):
    print(tag_freq_sorted[i])
    if(tag_freq_sorted[i][1]<=29):
        break

('life', 538)
('inspirational', 458)
('love', 353)
('humor', 266)
('positive', 167)
('books', 153)
('quotes', 122)
('living', 102)
('self', 99)
('reading', 96)
('inspiration', 87)
('of', 86)
('misattributed', 82)
('writing', 76)
('happiness', 69)
('philosophy', 68)
('wisdom', 66)
('friendship', 65)
('truth', 64)
('the', 53)
('and', 53)
('death', 51)
('jace', 51)
('thinking', 51)
('optimism', 48)
('romance', 47)
('hope', 46)
('women', 46)
('poetry', 45)
('affirmation', 41)
('lessons', 40)
('optimistic', 40)
('courage', 39)
('inspire', 38)
('inspiring', 38)
('religion', 36)
('attitude', 36)
('change', 34)
('fear', 34)
('pain', 34)
('funny', 34)
('music', 33)
('to', 33)
('relationships', 33)
('dreams', 32)
('science', 31)
('friends', 30)
('strength', 30)
('wayland', 30)
('god', 30)
('clary', 29)


## Removal of Stop Words in the Target Column

# Multi-Label Naive Bayes Classifier (Baseline Model)

# Recurrant Neural Network (RNN) Model

# Transformer Based Model?